### ISSUE TITLE
NumberFormatException // valid use?

### ISSUE THREAD

**Issue by blackfoxoh** [uniVocity/univocity-parsers#156]
hi

I'm receiving a NumberFormatException while parsing a File. A relevant sample of such a stacktrace looks like the following:
```Exception data: java.lang.NumberFormatException: multiple points
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:1138)
	at java.lang.Double.parseDouble(Double.java:521)
	at java.text.DigitList.getDouble(DigitList.java:162)
	at java.text.DecimalFormat.parse(DecimalFormat.java:1313)
	at java.text.SimpleDateFormat.subParse(SimpleDateFormat.java:1620)
	at java.text.SimpleDateFormat.parse(SimpleDateFormat.java:1343)
	at java.text.DateFormat.parse(DateFormat.java:346)
	at com.univocity.parsers.conversions.DateConversion.fromString(DateConversion.java:92)
	at com.univocity.parsers.conversions.DateConversion.fromString(DateConversion.java:36)
	at com.univocity.parsers.conversions.ObjectConversion.fromInput(ObjectConversion.java:69)
	at com.univocity.parsers.conversions.ObjectConversion.fromInput(ObjectConversion.java:29)
	at com.univocity.parsers.conversions.NullConversion.execute(NullConversion.java:66)
	at com.univocity.parsers.conversions.ObjectConversion.execute(ObjectConversion.java:58)
	at com.univocity.parsers.conversions.ObjectConversion.execute(ObjectConversion.java:29)
	at com.univocity.parsers.common.record.MetaData.convert(MetaData.java:47)
	at com.univocity.parsers.common.record.RecordMetaDataImpl.convert(RecordMetaDataImpl.java:281)
	at com.univocity.parsers.common.record.RecordMetaDataImpl.getObjectValue(RecordMetaDataImpl.java:354)
	at com.univocity.parsers.common.record.RecordImpl.getDate(RecordImpl.java:580)
	at <my own code - calling 'record.getDate(RGDATUM);'>
```

NumberFormatException seems not to be catched here but apart from that I think my true problem could be a concurrency-issue or something else really weird. 
The input-file has lines like the following, containing 3 columns:
```
0009999997;20160706;00004220057
0009999998;20160706;00004220058
0009999999;20160706;00004220059
...
```
I cheked the input with a texteditor: every line of the input conforms to the following regexp `[0-9]{10};[0-9]{8};[0-9]{11}\r\n`

My parser is defined the following way:
```java
      CsvParserSettings cps = new CsvParserSettings();
      cps.getFormat().setDelimiter(';');
      cps.setLineSeparatorDetectionEnabled(true);
      cps.setHeaders(PAKETNR, RGDATUM, RGNR);
      CsvParser cp = new CsvParser(cps);

      cp.beginParsing(file);
      cp.getRecordMetadata().convertFields(new DateConversion("yyyyMMdd")).add(RGDATUM);
```
(PAKETNR, RGDATUM, RGNR are String-constants)

The input-file has about 220000 lines which is not too much for the parser I think. More complex in my case is the processing afterwards therefore I process a bunch of records in parallel:
```java
      Record record;
      int batchSize = 5000;
      Collection<Record> bunchOfRecords = new ArrayList<Record>(batchSize);
      Collection<Future<Integer>> asyncResults = new HashSet<Future<Integer>>();
      long baseLineNr = 0;
      while ((record = cp.parseNextRecord()) != null) {
         bunchOfRecords.add(record);
         importet++;
         if (importet % batchSize == 0) { // process x records in parallel
            asyncResults.add(self.parseRecords(bunchOfRecords, baseLineNr)); // asynchronous call(!)
            bunchOfRecords = new ArrayList<Record>(batchSize);
            baseLineNr = cp.getContext().currentLine();
         }
      }
      // asynchronous call for the last records (less than batchSize)
      asyncResults.add(self.parseRecords(bunchOfRecords, baseLineNr));

      cp.stopParsing();
```
in the asynchronous processing-method I use the records the receive the three column-values per record by calling 
```java
record.getInt(PAKETNR);
record.getDate(RGDATUM);
record.getInt(RGNR);
```

Is it valid to use the parsed Record-objects in an asynchronous call?
In the NumberFormatException the error is detailed with "multiple points". This sounds like I would try to parse a value like '2016.07.06' to a number. I just guess (!) this could be a concurrency-issue as this looks like a already parsed date is getting parsed again (compare the date-conversion for the second column).

Should records be thread-safe? I thought about switching to a parallel-parser as the documentation says but actually I don't want nor need the parsing to be in parallel as this is really fast enough...

I'm looking forward for your estimation on this

regards

**Comment by jbax** [comment]
Pretty sure it's happening because the [SimpleDataFormat](https://github.com/uniVocity/univocity-parsers/blob/e34a149077147de7cafcafbbaf8a4310e6297584/src/main/java/com/univocity/parsers/conversions/DateConversion.java#L38) instance is reused instead of recreated for each individual call.

This class is not thread-safe hence you get this error. This could be solved with the use of a `ThreadLocal` to keep the date format instances, but I'll have to check evaluate the performance impact. If it gets too bad I will probably add a `record.getDateSync(RGDATUM)` or something like that to parse dates concurrently safely.

**Comment by blackfoxoh** [comment]
that sounds applicable... Thanks for pointing me this out - as of my special usecase I could dismiss the dateconversion for this field and convert it on my own in a threadsafe manner. Of course, if a solution with `ThreadLocal` would be available soon that would be great as well.

As I've seen you have a "java 8"-label for some other issues: for a future release using java 8 features you might want to consider `java.time.format.DateTimeFormatter`.

Thanks again for your (as usual) ultrafast response!

**Comment by jbax** [comment]
Fixed and released a 2.5.0 SNAPSHOT build to include the adjustment.
